doc/guides/zephyr_cmake_package.rst

```diff
@@ -0,0 +1,418 @@
+     builds. To change the Zephyr package, as example to test the application using environment
+     ``ZEPHYR_BASE`` then it is neccessary to do a pristine build first
+     (See :ref:`application_rebuild`).
+
+Zephyr CMake Package Version (Zephyr Freestanding Application)
+**************************************************************
+
+When writing an out-of-tree application then it is possible to specify a Zephyr version number
+``x.y.z`` that must be used in order to build the application.
+
+For example:
+
+  .. code-block:: cmake
+
+     cmake_minimum_required(VERSION 3.13.1)
+     find_package(Zephyr 2.2.0)
+     project(app)
+
+will require ``app`` to be built with Zephyr 2.2.0 as minimum.
+CMake will search all exported candidates to find a Zephyr installation which matches this version
+criteria.
+
+Thus it is possible to have multiple Zephyr installations and have CMake automatically select
+between them based on the version number provided, see `CMake Package Version`_ for details.
+
+For example:
+
+.. code-block:: none
+
+   <projects>/zephyr-workspace-2.a
+   └── zephyr                       (Exported to CMake)
+
+   <projects>/zephyr-workspace-2.b
+   └── zephyr                       (Exported to CMake)
+
+   <home>/app
+   ├── CMakeLists.txt
+   ├── prj.conf
+   └── src
+       └── main.c
+
+in this case, there are two released versions of Zephyr installed at their own workspaces.
+Workspace 2.a and 2.b, corresponding to the Zephyr version.
+
+To ensure ``app`` is built with minimum version ``2.a`` the following ``find_package``
+syntax may be used:
+
+.. code-block:: cmake
+
+   cmake_minimum_required(VERSION 3.13.1)
+   find_package(Zephyr 2.a)
+   project(app)
+
+
+Note that both ``2.a`` and ``2.b`` fulfill this requirement.
+
+CMake also supports the keyword ``EXACT``, to ensure an exact version is used, if that is required.
+In this case, the application CMakeLists.txt could be written as:
+
+.. code-block:: cmake
+
+   cmake_minimum_required(VERSION 3.13.1)
+   find_package(Zephyr 2.a EXACT)
+   project(app)
+
+In case no Zephyr is found which satisfies the version required, as example, the application specifies
+
+.. code-block:: cmake
+
+   cmake_minimum_required(VERSION 3.13.1)
+   find_package(Zephyr 2.z)
+   project(app)
+
+then an error similar to below will be printed::
+
+  Could not find a configuration file for package "Zephyr" that is compatible
+  with requested version "2.z".
+
+  The following configuration files were considered but not accepted:
+
+    <projects>/zephyr-workspace-2.a/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake, version: 2.a.0
+    <projects>/zephyr-workspace-2.b/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake, version: 2.b.0
+
+
+.. note:: It is also possible to specify a version number for both Zephyr repository applications
+          and Zephyr workspace applications. This can be used to still ensure the correct Zephyr
```

Comment:
Does this mean it won't search exported Zephyrs if the version doesn't match?

Reply:
It will be searched, but not used, and thus give an error such as:
```
Could not find a configuration file for package "Zephyr" that is compatible
with requested version "2.z".

The following configuration files were considered but not accepted:

  <projects>/zephyr-workspace-2.a/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake, version: 2.a.0
```
if user specifies that `2.z` is required but the Zephyr version is only `2.a`.

This could happen for someone who is working on a not-commited app (dirty-repo), and then check-outs and older Zephyr for other reasons.
Switching back to working on the not-commited app without checking out the new Zephyr version.

Reply:
What if there's an exported 2.z. Will the app use that instead?

Reply:
In a freestanding app, yes.

If you are working in a Zephyr repository app (let's say Zephyr 2.a) and specifies 2z, and there is a Zephyr 2.z else-where in the system, then no.
Cause this would mean you would be using a Zephyr OUTSIDE your current Zephyr repo.
This would most likely be a user-error, and if we decided to use the 2.z else-where, I believe we might confuse users.

If a user in this case really wants to use the 2.z version, then he can specify `ZEPHYR_BASE=/<path>/<to>/zephyr (2.z)`, cause `ZEPHYR_BASE` can override everything.